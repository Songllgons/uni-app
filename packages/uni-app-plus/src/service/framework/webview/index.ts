import { ON_WEBVIEW_READY } from '../../../constants'
import { createNVueWebview } from './nvue'
import { getWebviewId } from './utils'
import { getPreloadWebview } from './preload'

export * from './init'

export * from './preload'

export interface CreateWebviewOptions {
  path: string
  query: Record<string, string>
  routeOptions: UniApp.UniRoute
  webviewExtras?: Record<string, any>
}

export function createWebview(options: CreateWebviewOptions) {
  if (options.routeOptions.meta.isNVue) {
    return createNVueWebview(options)
  }
  if (getWebviewId() === 2) {
    // 如果首页非 nvue，则直接返回 Launch Webview
    return plus.webview.getLaunchWebview()
  }
  return getPreloadWebview()
}

export function onWebviewReady(
  pageId: string,
  callback: (...args: any[]) => void
) {
  UniServiceJSBridge.once(ON_WEBVIEW_READY + '.' + pageId, callback)
}
