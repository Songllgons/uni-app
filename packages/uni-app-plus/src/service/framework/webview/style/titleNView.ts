import { isArray } from '@vue/shared'
import { ON_NAVIGATION_BAR_BUTTON_TAP, formatLog } from '@dcloudio/uni-shared'
import { initNavigationBarI18n, invokeHook } from '@dcloudio/uni-core'
export function initTitleNView(
  webviewStyle: PlusWebviewWebviewStyles,
  routeMeta: UniApp.PageRouteMeta
) {
  const { navigationBar } = routeMeta
  if (navigationBar.style === 'custom') {
    return false
  }
  let autoBackButton = true
  if (routeMeta.isQuit) {
    autoBackButton = false
  }
  const titleNView: PlusWebviewWebviewTitleNViewStyles = {
    autoBackButton,
  }
  Object.keys(navigationBar).forEach((name) => {
    const value = navigationBar[name as keyof UniApp.PageNavigationBar]
    if (name === 'titleImage' && value) {
      titleNView.tags = createTitleImageTags(value as string)
    } else if (name === 'buttons' && isArray(value)) {
      titleNView.buttons = (value as UniApp.PageNavigationBar['buttons'])!.map(
        (button, index) => {
          ;(button as any).onclick = createTitleNViewBtnClick(index)
          return button
        }
      )
    } else {
      titleNView[name as keyof PlusWebviewWebviewTitleNViewStyles] =
        value as any
    }
  })

  webviewStyle.titleNView = initTitleNViewI18n(titleNView, routeMeta)
}

function initTitleNViewI18n(
  titleNView: PlusWebviewWebviewTitleNViewStyles,
  routeMeta: UniApp.PageRouteMeta
) {
  const i18nResult = initNavigationBarI18n(titleNView)
  if (!i18nResult) {
    return titleNView
  }
  const [titleTextI18n, searchInputPlaceholderI18n] = i18nResult
  if (titleTextI18n || searchInputPlaceholderI18n) {
    uni.onLocaleChange(() => {
      const webview = plus.webview.getWebviewById(routeMeta.id + '')
      if (!webview) {
        return
      }
      const newTitleNView: PlusWebviewWebviewTitleNViewStyles = {}
      if (titleTextI18n) {
        newTitleNView.titleText = titleNView.titleText
      }
      if (searchInputPlaceholderI18n) {
        newTitleNView.searchInput = {
          placeholder: titleNView.searchInput!.placeholder,
        }
      }
      if (__DEV__) {
        console.log(formatLog('updateWebview', webview.id, newTitleNView))
      }
      webview.setStyle({
        titleNView: newTitleNView,
      })
    })
  }
  return titleNView
}

function createTitleImageTags(titleImage: string) {
  return [
    {
      tag: 'img' as 'img',
      src: titleImage,
      position: {
        left: 'auto',
        top: 'auto',
        width: 'auto',
        height: '26px',
      },
    },
  ]
}

function createTitleNViewBtnClick(index: number) {
  return function onClick(btn: UniApp.PageNavigationBarButton) {
    ;(btn as any).index = index
    invokeHook(ON_NAVIGATION_BAR_BUTTON_TAP, btn)
  }
}
