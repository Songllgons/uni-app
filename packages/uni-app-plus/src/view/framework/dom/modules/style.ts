import { hyphenate, isArray, isString } from '@vue/shared'
import { normalizeStyleName, normalizeStyleValue } from '../../../utils'
import type { UniCustomElement } from '../components'

export function patchStyle(
  el: UniCustomElement,
  value: string | Record<string, any>
) {
  const style = (el as HTMLElement).style
  if (isString(value)) {
    if (value === '') {
      el.removeAttribute('style')
    } else {
      // TODO display
      style.cssText = normalizeStyleValue(value)
    }
  } else {
    for (const key in value) {
      setStyle(style, key, value[key])
    }
  }

  const { __wxsStyle } = el
  if (__wxsStyle) {
    for (const key in __wxsStyle) {
      setStyle(style, key, __wxsStyle[key] as string)
    }
  }
}

const importantRE = /\s*!important$/

function setStyle(
  style: CSSStyleDeclaration,
  name: string,
  val: string | string[]
) {
  if (isArray(val)) {
    val.forEach((v) => setStyle(style, name, v))
  } else {
    val = normalizeStyleValue(val)
    if (name.startsWith('--')) {
      // custom property definition
      style.setProperty(name, val)
    } else {
      const prefixed = normalizeStyleName(style, name)
      if (importantRE.test(val)) {
        // !important
        style.setProperty(
          hyphenate(prefixed),
          val.replace(importantRE, ''),
          'important'
        )
      } else {
        style[prefixed as any] = val
      }
    }
  }
}
