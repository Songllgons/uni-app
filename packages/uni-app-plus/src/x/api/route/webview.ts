import type { IPage } from '@dcloudio/uni-app-x/types/native'

export function showWebview(
  nPage: IPage,
  animationType: string,
  animationDuration: number,
  showCallback?: () => void
) {
  nPage.show(
    new Map<string, any>([
      ['animationType', animationType],
      ['animationDuration', animationDuration],
    ]),
    showCallback
  )
}

export function closeWebview(
  nPage: IPage,
  animationType: string,
  animationDuration?: number,
  callback?: () => void
) {
  const options = new Map<string, any>([['animationType', animationType]])
  if (typeof animationDuration === 'number') {
    options.set('animationDuration', animationDuration)
  }
  nPage.close(options, callback)
}
