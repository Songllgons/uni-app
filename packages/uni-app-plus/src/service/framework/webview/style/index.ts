import { initNVue } from './nvue'
import { initBackgroundColor } from './backgroundColor'
import { initPopGesture } from './popGesture'
import { initPullToRefresh } from './pullToRefresh'
import { initTitleNView } from './titleNView'
import type { DebugRefresh, InitUniPageUrl } from '../utils'

export function parseWebviewStyle(
  path: string,
  routeMeta: UniApp.PageRouteMeta,
  webview: { id: string }
): PlusWebviewWebviewStyles & {
  uniPageUrl?: InitUniPageUrl
  debugRefresh?: DebugRefresh
  isTab?: boolean
  locale?: string
} {
  const webviewStyle: PlusWebviewWebviewStyles = {
    bounce: 'vertical',
  }

  Object.keys(routeMeta).forEach((name) => {
    if (WEBVIEW_STYLE_BLACKLIST.indexOf(name) === -1) {
      webviewStyle[name as keyof PlusWebviewWebviewStyles] =
        routeMeta[name as keyof UniApp.PageRouteMeta]
    }
  })
  if (webview.id !== '1') {
    // 首页 nvue 已经在 manifest.json 中设置了 uniNView，不能再次设置，否则会二次加载
    initNVue(webviewStyle, routeMeta, path)
  }
  initPopGesture(webviewStyle, routeMeta)
  initBackgroundColor(webviewStyle, routeMeta)
  initTitleNView(webviewStyle, routeMeta)
  initPullToRefresh(webviewStyle, routeMeta)

  return webviewStyle
}

const WEBVIEW_STYLE_BLACKLIST = [
  'id',
  'route',
  'isNVue',
  'isQuit',
  'isEntry',
  'isTabBar',
  'tabBarIndex',
  'windowTop',
  'topWindow',
  'leftWindow',
  'rightWindow',
  'maxWidth',
  'usingComponents',
  'disableScroll',
  'enablePullDownRefresh',
  'navigationBar',
  'pullToRefresh',
  'onReachBottomDistance',
  'pageOrientation',
  'backgroundColor',
]
