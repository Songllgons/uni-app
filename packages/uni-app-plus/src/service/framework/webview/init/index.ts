import { initWebviewEvent } from './event'
import { initWebviewStyle } from './style'
import { initSubNVues } from './subNVues'

export function initWebview(
  webview: PlusWebviewWebviewObject,
  path: string,
  query: Record<string, any>,
  routeMeta: UniApp.PageRouteMeta
) {
  // 首页或非 nvue 页面
  if (webview.id === '1' || !routeMeta.isNVue) {
    initWebviewStyle(webview, path, query, routeMeta)
  }
  initSubNVues(webview, path, routeMeta)
  initWebviewEvent(webview)
}
