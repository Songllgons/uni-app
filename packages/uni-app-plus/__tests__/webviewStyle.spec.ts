import { parseWebviewStyle } from '../src/service/framework/webview/style'
function initDefaultUniConfig() {
  return JSON.parse(
    JSON.stringify({
      globalStyle: {
        navigationBar: {},
      },
    })
  )
}
beforeAll(() => {
  global.__uniConfig = initDefaultUniConfig()
})
const defaultPath = '/pages/index/index'
const defaultRoute = 'pages/index/index'
describe('webviewStyle', () => {
  test('basic', () => {
    expect(
      parseWebviewStyle(
        defaultPath,
        {
          route: defaultRoute,
          navigationBar: {},
        },
        { id: '2' } as any
      )
    ).toMatchSnapshot()
  })
})
