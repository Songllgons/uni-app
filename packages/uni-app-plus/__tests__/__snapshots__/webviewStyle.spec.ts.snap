// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webviewStyle basic 1`] = `
{
  "bounce": "vertical",
  "titleNView": {
    "autoBackButton": true,
  },
}
`;
